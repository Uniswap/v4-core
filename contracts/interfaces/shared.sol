// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.10;

/// @notice Represents a change in the pool's balance of token0 and token1.
struct BalanceDelta {
    int256 amount0;
    int256 amount1;
}
