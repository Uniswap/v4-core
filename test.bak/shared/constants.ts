export const MIN_TICK = -887272
export const MAX_TICK = 887272
export const MAX_TICK_SPACING = 32767
export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
