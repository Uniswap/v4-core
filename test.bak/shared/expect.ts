import { expect, use } from 'chai'
import { solidity } from 'ethereum-waffle'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'

use(solidity)
use(jestSnapshotPlugin())

export { expect }
