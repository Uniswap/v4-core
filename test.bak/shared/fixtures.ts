import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { TestERC20 } from '../../typechain/TestERC20'

interface TokensFixture {
  currency0: TestERC20
  currency1: TestERC20
  token2: TestERC20
}

export async function tokensFixture(): Promise<TokensFixture> {
  const tokenFactory = await ethers.getContractFactory('TestERC20')
  const tokenA = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenB = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenC = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20

  const [currency0, currency1, token2] = [tokenA, tokenB, tokenC].sort((tokenA, tokenB) =>
    tokenA.address.toLowerCase() < tokenB.address.toLowerCase() ? -1 : 1
  )

  return { currency0, currency1, token2 }
}
