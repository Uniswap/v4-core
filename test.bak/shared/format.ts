import { Decimal } from 'decimal.js'
import { BigNumberish } from 'ethers'

export function formatTokenAmount(num: BigNumberish): string {
  return new Decimal(num.toString()).dividedBy(new Decimal(10).pow(18)).toPrecision(5)
}

export function formatPrice(price: BigNumberish): string {
  return new Decimal(price.toString()).dividedBy(new Decimal(2).pow(96)).pow(2).toPrecision(5)
}
