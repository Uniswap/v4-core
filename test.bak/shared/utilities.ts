import bn from 'bignumber.js'
import { BigNumber, BigNumberish, Contract, ContractTransaction, utils, Wallet } from 'ethers'
import { ethers } from 'hardhat'

export const MaxUint128 = BigNumber.from(2).pow(128).sub(1)

export const getMinTick = (tickSpacing: number) => Math.ceil(-887272 / tickSpacing) * tickSpacing
export const getMaxTick = (tickSpacing: number) => Math.floor(887272 / tickSpacing) * tickSpacing

export const MIN_SQRT_RATIO = BigNumber.from('4295128739')
export const MAX_SQRT_RATIO = BigNumber.from('1461446703485210103287273052203988822378723970342')

export enum FeeAmount {
  LOW = 500,
  MEDIUM = 3000,
  HIGH = 10000,
}

export const TICK_SPACINGS: { [amount in FeeAmount]: number } = {
  [FeeAmount.LOW]: 10,
  [FeeAmount.MEDIUM]: 60,
  [FeeAmount.HIGH]: 200,
}

export function expandTo18Decimals(n: number): BigNumber {
  return BigNumber.from(n).mul(BigNumber.from(10).pow(18))
}

bn.config({ EXPONENTIAL_AT: 999999, DECIMAL_PLACES: 40 })

// returns the sqrt price as a 64x96
export function encodeSqrtPriceX96(reserve1: BigNumberish, reserve0: BigNumberish): BigNumber {
  return BigNumber.from(
    new bn(reserve1.toString())
      .div(reserve0.toString())
      .sqrt()
      .multipliedBy(new bn(2).pow(96))
      .integerValue(3)
      .toString()
  )
}

export function getPositionKey(address: string, lowerTick: number, upperTick: number): string {
  return utils.keccak256(utils.solidityPack(['address', 'int24', 'int24'], [address, lowerTick, upperTick]))
}

export function getPoolId({
  currency0,
  currency1,
  fee,
  tickSpacing,
  hooks,
}: {
  currency0: string | Contract
  currency1: string | Contract
  fee: number
  tickSpacing: number
  hooks: string | Contract
}): string {
  return utils.keccak256(
    utils.defaultAbiCoder.encode(
      ['address', 'address', 'uint24', 'int24', 'address'],
      [
        typeof currency0 === 'string' ? currency0 : currency0.address,
        typeof currency1 === 'string' ? currency1 : currency1.address,
        fee,
        tickSpacing,
        typeof hooks === 'string' ? hooks : hooks.address,
      ]
    )
  )
}

export type SwapFunction = (
  amount: BigNumberish,
  to: Wallet | string,
  sqrtPriceLimitX96?: BigNumberish
) => Promise<ContractTransaction>
export type SwapToPriceFunction = (sqrtPriceX96: BigNumberish, to: Wallet | string) => Promise<ContractTransaction>
export type ModifyPositionFunction = (
  tickLower: BigNumberish,
  tickUpper: BigNumberish,
  liquidityDelta: BigNumberish
) => Promise<ContractTransaction>
export type DonateFunction = (amount0: BigNumberish, amount1: BigNumberish) => Promise<ContractTransaction>

interface HookMask {
  beforeInitialize: boolean
  afterInitialize: boolean
  beforeModifyPosition: boolean
  afterModifyPosition: boolean
  beforeSwap: boolean
  afterSwap: boolean
  beforeDonate: boolean
  afterDonate: boolean
}

/**
 * Creates a 20 byte mask for the given hook configuration
 */
export function createHookMask({
  beforeInitialize,
  afterInitialize,
  beforeModifyPosition,
  afterModifyPosition,
  beforeSwap,
  afterSwap,
  beforeDonate,
  afterDonate,
}: HookMask): string {
  let result: BigNumber = BigNumber.from(0)
  if (beforeInitialize) result = result.add(BigNumber.from(1).shl(159))
  if (afterInitialize) result = result.add(BigNumber.from(1).shl(158))
  if (beforeModifyPosition) result = result.add(BigNumber.from(1).shl(157))
  if (afterModifyPosition) result = result.add(BigNumber.from(1).shl(156))
  if (beforeSwap) result = result.add(BigNumber.from(1).shl(155))
  if (afterSwap) result = result.add(BigNumber.from(1).shl(154))
  if (beforeDonate) result = result.add(BigNumber.from(1).shl(153))
  if (afterDonate) result = result.add(BigNumber.from(1).shl(152))
  return utils.hexZeroPad(result.toHexString(), 20)
}

/**
 * Returns a wallet whose first transaction will create a contract satisfying the leading
 * bytes required by hookMask. If provided, the mnemonic argument short-circuits our search
 * to save time.
 */
export function getWalletForDeployingHookMask(hookMask: HookMask, mnemonic?: string): [Wallet, string] {
  const startingString = createHookMask(hookMask).slice(0, 4)
  let wallet: Wallet = mnemonic ? ethers.Wallet.fromMnemonic(mnemonic) : ethers.Wallet.createRandom()
  let contractAddress: string | undefined

  while (contractAddress === undefined) {
    const prospectiveContractAddress = utils.getContractAddress({ from: wallet.address, nonce: 0 })
    if (prospectiveContractAddress.slice(0, 4).toLowerCase() === startingString) {
      contractAddress = prospectiveContractAddress
    } else {
      // if, for whatever reason, we generate a bad address but a mnemonic was provided,
      // it's stale and we surface the error
      if (mnemonic) throw Error('Stale mnemonic')
      wallet = ethers.Wallet.createRandom()
    }
  }

  return [wallet, contractAddress]
}
