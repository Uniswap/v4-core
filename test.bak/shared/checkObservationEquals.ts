import { BigNumber, BigNumberish } from 'ethers'
import { expect } from './expect'

// helper function because we cannot do a simple deep equals with the
// observation result object returned from ethers because it extends array
export default function checkObservationEquals(
  {
    tickCumulative,
    blockTimestamp,
    initialized,
    secondsPerLiquidityCumulativeX128,
  }: {
    tickCumulative: BigNumber
    secondsPerLiquidityCumulativeX128: BigNumber
    initialized: boolean
    blockTimestamp: number
  },
  expected: {
    tickCumulative: BigNumberish
    secondsPerLiquidityCumulativeX128: BigNumberish
    initialized: boolean
    blockTimestamp: number
  }
) {
  expect(
    {
      initialized,
      blockTimestamp,
      tickCumulative: tickCumulative.toString(),
      secondsPerLiquidityCumulativeX128: secondsPerLiquidityCumulativeX128.toString(),
    },
    `observation is equivalent`
  ).to.deep.eq({
    ...expected,
    tickCumulative: expected.tickCumulative.toString(),
    secondsPerLiquidityCumulativeX128: expected.secondsPerLiquidityCumulativeX128.toString(),
  })
}
