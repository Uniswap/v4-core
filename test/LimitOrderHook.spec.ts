import { Wallet } from 'ethers'
import hre, { ethers, waffle } from 'hardhat'
import { LimitOrderHook, PoolManager, TestERC20 } from '../typechain'
import { expect } from './shared/expect'
import { tokensFixture } from './shared/fixtures'
import { encodeSqrtPriceX96, FeeAmount, getPoolId, getWalletForDeployingHookMask } from './shared/utilities'

const { constants } = ethers

const createFixtureLoader = waffle.createFixtureLoader

describe('LimitOrderHooks', () => {
  let wallet: Wallet, other: Wallet

  let tokens: { token0: TestERC20; token1: TestERC20; token2: TestERC20 }
  let manager: PoolManager
  let limitOrderHook: LimitOrderHook

  const fixture = async () => {
    const tokens = await tokensFixture()

    const poolManagerFactory = await ethers.getContractFactory('PoolManager')
    const limitOrderHookFactory = await ethers.getContractFactory('LimitOrderHook')

    const manager = (await poolManagerFactory.deploy()) as PoolManager

    // find a deployer that will generate a suitable hooks address
    const [hookDeployer, hookAddress] = getWalletForDeployingHookMask(
      {
        beforeInitialize: false,
        afterInitialize: true,
        beforeModifyPosition: false,
        afterModifyPosition: false,
        beforeSwap: false,
        afterSwap: true,
        beforeDonate: false,
        afterDonate: false,
      },
      'test onion mountain stove water behind cloud street robot salad load join'
    )

    ;[wallet] = await (ethers as any).getSigners()
    await wallet.sendTransaction({ to: hookDeployer.address, value: ethers.utils.parseEther('1') })

    // deploy the hook and make a contract instance
    await hookDeployer
      .connect(hre.ethers.provider)
      .sendTransaction(limitOrderHookFactory.getDeployTransaction(manager.address))
    const limitOrderHook = limitOrderHookFactory.attach(hookAddress) as LimitOrderHook

    const result = {
      manager,
      tokens,
      limitOrderHook,
    }

    for (const token of [tokens.token0, tokens.token1, tokens.token2]) {
      for (const spender of [limitOrderHook]) {
        await token.connect(wallet).approve(spender.address, constants.MaxUint256)
      }
    }

    return result
  }

  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    ;[wallet, other] = await (ethers as any).getSigners()

    loadFixture = createFixtureLoader([wallet, other])
  })

  beforeEach('deploy fixture', async () => {
    ;({ manager, tokens, limitOrderHook } = await loadFixture(fixture))
  })

  it('bytecode size', async () => {
    expect(((await waffle.provider.getCode(limitOrderHook.address)).length - 2) / 2).to.matchSnapshot()
  })

  describe('#initialize', async () => {
    it('initialize a pool with limit order hook', async () => {
      await manager.initialize(
        {
          token0: tokens.token0.address,
          token1: tokens.token1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: limitOrderHook.address,
        },
        encodeSqrtPriceX96(1, 1)
      )

      const {
        slot0: { sqrtPriceX96 },
      } = await manager.pools(
        getPoolId({
          token0: tokens.token0.address,
          token1: tokens.token1.address,
          fee: FeeAmount.MEDIUM,
          tickSpacing: 60,
          hooks: limitOrderHook.address,
        })
      )
      expect(sqrtPriceX96).to.eq(encodeSqrtPriceX96(1, 1))
    })
  })
})
