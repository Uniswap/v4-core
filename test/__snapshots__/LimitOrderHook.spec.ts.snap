// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitOrderHooks bytecode size 1`] = `13038`;
